pkg/agent/storage/storage.go

```diff
@@ -0,0 +1,287 @@
+package storage
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/spiffe/spire/pkg/common/diskutil"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+)
+
+var (
+	ErrNotCached = errors.New("not cached")
+)
+
+type Storage interface {
+	// LoadSVID loads the SVID from storage. Returns ErrNotCached if the SVID
+	// does not exist in the cache.
+	LoadSVID() ([]*x509.Certificate, error)
+
+	// StoreSVID stores the SVID.
+	StoreSVID(certs []*x509.Certificate) error
+
+	// DeleteSVID deletes the SVID.
+	DeleteSVID() error
+
+	// LoadBundle loads the SVID from storage. Returns ErrNotCached if the
+	// bundle does not exist in the cache.
+	LoadBundle() ([]*x509.Certificate, error)
+
+	// StoreBundle stores the bundle.
+	StoreBundle(certs []*x509.Certificate) error
+}
+
+func Open(dir string) (Storage, error) {
+	legacySVID, legacySVIDTime, err := loadLegacySVID(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	legacyBundle, legacyBundleTime, err := loadLegacyBundle(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	data, dataTime, err := loadData(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
```

Comment:
I think that it would be good if we can provide some context to enrich the errors returned, otherwise it may be difficult for the operators to identify under what operation occurred the errors.
For example, in this case an error when unmarshaling data in the loadData function will be returned as is and the operator will have no clue that this happened under the context of loading agent data fro the json file. This applies in general across this implementation.

Reply:
Sounds good. I added context to all of the failure paths in the lower level functions. Let me know if you think I missed something.